### PR TITLE
DevCenter DataPlane v2023-04-01 stable SDK from TypeSpec

### PR DESCRIPTION
**Pull request by drielenr**
# Description

We currently have a beta SDK targeting DevCenter 2023-04-01 API, and this PR is to have it released as a stable SDK - our first stable release, targeting same API. 

Both SDKs (the last one released and this one) were generated from TypeSpec. The main change introduced was add model generation and updates on tests and samples to reflect that addition. 

Arch board meeting occurred but APIView is not yet approved. Their requests have been addressed in this PR.

# All SDK Contribution checklist:
- [] **The pull request does not introduce breaking changes**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### Pipeline runs

Dry runs on the release pipeline:
https://dev.azure.com/azure-sdk/internal/_build?definitionId=5622&_a=summary

### List of tests

DevBox
 - [x] get_pool
 - [x] list_pools
 - [x] get_dev_box
 - [x] get_remote_connection
 - [x] get_action
 - [x] list_actions
 - [x] skip_action
 - [x] delay_action
 - [x] delay_all_actions
 - [x] get_schedule
 - [x] list_schedules
 - [x] list_dev_boxes
 - [x] list_all_dev_boxes
 - [x] list_all_dev_boxes_by_user
 - [x] begin_create_dev_box
 - [x] begin_delete_dev_box
 - [x] begin_start_dev_box
 - [x] begin_stop_dev_box
 - [x] begin_restart_dev_box

DevCenter
 - [x] get_project
 - [x] list_projects

Environments
 - [x] begin_delete_environment
 - [x] begin_create_or_update_environment
 - [x] list_environment_types
 - [x] list_catalogs
 - [x] list_environments
 - [x] list_all_environments
 - [x] get_environment_definition
 - [x] list_environment_definitions
 - [x]  list_environment_definitions_by_catalog
 - [x] get_catalog
 - [x] get_environment

Also updated samples.

### CODE REVIEW

File: sdk/devcenter/azure-developer-devcenter/README.md
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/README.md:71]: These are causing your build errors, there's a misspelling also make sure it matches the sections you're referencing

File: sdk/devcenter/azure-developer-devcenter/azure/developer/devcenter/_operations/_operations.py
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/azure/developer/devcenter/_operations/_operations.py:3025]: Just checking is this the expected return type?

File: sdk/devcenter/azure-developer-devcenter/azure/developer/devcenter/_operations/_operations.py
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/azure/developer/devcenter/_operations/_operations.py:3885]: This operation didnt have a return type?

File: sdk/devcenter/azure-developer-devcenter/samples/create_devbox_sample.py
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/samples/create_devbox_sample.py:44]: You can move this to its own file since you also create a client in the function below and creating the dev box is the more interesting part of what we're doing here.

File: sdk/devcenter/azure-developer-devcenter/samples/create_environment_sample.py
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/samples/create_environment_sample.py:121]:
```suggestion
    environment_poller = client.begin_create_or_update_environment(
```

File: sdk/devcenter/azure-developer-devcenter/samples/create_environment_sample.py
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/samples/create_environment_sample.py:129]:
```suggestion
    delete_poller = client.begin_delete_environment(target_project_name, "me", environment_name)
```

File: sdk/devcenter/azure-developer-devcenter/samples/devbox_action_sample copy.py
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/samples/devbox_action_sample copy.py:1]: Let's fix this file name.

File: sdk/devcenter/azure-developer-devcenter/sdk-generation.ps1
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/sdk-generation.ps1:1]: Not sure if we should be checking this file in, will check on it

File: sdk/devcenter/azure-developer-devcenter/tests/test_devcenter_operations.py
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/tests/test_devcenter_operations.py:50]: Will this value always be 1 or should this check allow for more than 1 project by checking > 1?

**Review by catalinaperalta — COMMENTED** [review]
We need async versions of all the samples. Same for tests, we need to have async tests. Also, please update samples to include the AZURE_TENANT_ID for authentication.

File: sdk/devcenter/azure-developer-devcenter/azure/developer/devcenter/_operations/_operations.py
Comment (drielenr) [reply sdk/devcenter/azure-developer-devcenter/azure/developer/devcenter/_operations/_operations.py:3025]: Yes, this is correct

File: sdk/devcenter/azure-developer-devcenter/azure/developer/devcenter/_operations/_operations.py
Comment (drielenr) [reply sdk/devcenter/azure-developer-devcenter/azure/developer/devcenter/_operations/_operations.py:3885]: No,it does not. You can see in the record the response body is null https://github.com/Azure/azure-sdk-assets/blob/95190a8afc4f186def04729c747b3a129d96e76c/net/sdk/devcenter/Azure.Developer.DevCenter/tests/SessionRecords/DevBoxesClientTests/SkipActionAndDeleteDevBoxSucceeds.json#L69C1-L96C7

File: sdk/devcenter/azure-developer-devcenter/tests/test_devcenter_operations.py
Comment (drielenr) [reply sdk/devcenter/azure-developer-devcenter/tests/test_devcenter_operations.py:50]: We set up the resources so this is always 1

File: sdk/devcenter/azure-developer-devcenter/sdk-generation.ps1
Comment (drielenr) [reply sdk/devcenter/azure-developer-devcenter/sdk-generation.ps1:1]: I don't see any specific problem in have it in here, and it would make the generation a little easier for whoever is updating the sdks. But we can remove it too if you think it's inappropriate..

File: sdk/devcenter/azure-developer-devcenter/README.md
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/README.md:28]: If this can be retrieved from the portal somewhere, it would be good to mention that here.

File: sdk/devcenter/azure-developer-devcenter/samples/create_client_sample.py
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/samples/create_client_sample.py:1]: Looking at this separately now, I think it's better to just delete this sample and add the snippet blocks `# [START create_dev_center_client]` in one of your existing samples.

**Review by catalinaperalta — COMMENTED** [review]
Looks like we're still missing async samples and tests, we need to have some async coverage before releasing the stable SDK.

**Comment by leti367** [comment]
> Looks like we're still missing async samples and tests, we need to have some async coverage before releasing the stable SDK.

Added the async tests and samples in the latest iteration.

File: sdk/devcenter/azure-developer-devcenter/samples/async_samples/dev_box_restart_async_sample.py
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/samples/async_samples/dev_box_restart_async_sample.py:82]: Looks like we're missing an `async with client:` line in this sample...I'd add one before you start making the client calls from this line onward

File: sdk/devcenter/azure-developer-devcenter/samples/async_samples/dev_box_create_async_sample.py
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/samples/async_samples/dev_box_create_async_sample.py:91]: Missing async with client , you can add it before this line.

File: sdk/devcenter/azure-developer-devcenter/samples/async_samples/dev_box_action_async_sample.py
Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/samples/async_samples/dev_box_action_async_sample.py:91]: Needs async with client

Comment (catalinaperalta) [inline sdk/devcenter/azure-developer-devcenter/samples/async_samples/dev_box_action_async_sample.py:91]:
```suggestion
    async with client:
         await client.skip_dev_box_action(target_dev_box.project_name, "me", target_dev_box.name, "schedule-default")
```

**Review by catalinaperalta — COMMENTED** [review]
Looking good, we just need to get CI passing and add a couple of fixes in the async samples

**Review by catalinaperalta — APPROVED** [review]

**Review by mccoyp — COMMENTED** [review]
I would recommend rebasing this PR with the latest `main` branch if possible. There were a number of recent changes to test sanitizing, which could affect playback tests once this is merged. I see that the [test proxy version being used](https://github.com/Azure/azure-sdk-for-python/blob/drielene/dev-center-stable-sdk/eng/target_proxy_version.txt) in the development branch is a month behind the latest, meaning that the bulk of changes aren't being used at the moment.

EDIT: Any test issues could alternatively be addressed after merging, since rebasing such a large PR would probably be very difficult -- I definitely wouldn't block this PR on it.

**Comment by leti367** [comment]
> I would recommend rebasing this PR with the latest `main` branch if possible. There were a number of recent changes to test sanitizing, which could affect playback tests once this is merged. I see that the [test proxy version being used](https://github.com/Azure/azure-sdk-for-python/blob/drielene/dev-center-stable-sdk/eng/target_proxy_version.txt) in the development branch is a month behind the latest, meaning that the bulk of changes aren't being used at the moment.
> 
> EDIT: Any test issues could alternatively be addressed after merging, since rebasing such a large PR would probably be very difficult -- I definitely wouldn't block this PR on it.

Got it! I will rebase it.

**Review by mccoyp — APPROVED** [review]

File: sdk/devcenter/azure-developer-devcenter/tests/conftest.py
Comment (mccoyp) [inline sdk/devcenter/azure-developer-devcenter/tests/conftest.py:9]: I think the tests are taking so long in playback because `time.sleep` and `asyncio.sleep` aren't being patched to get bypassed (these are used when waiting between polling operations). If you import `is_live` from `devtools_testutils` and `mock` from `unittest`, try adding the following code to the bottom of this file:
```python
@pytest.fixture(scope="session", autouse=True)
def patch_async_sleep():
    async def immediate_return(_):
        return

    if not is_live():
        with mock.patch("asyncio.sleep", immediate_return):
            yield

    else:
        yield


@pytest.fixture(scope="session", autouse=True)
def patch_sleep():
    def immediate_return(_):
        return

    if not is_live():
        with mock.patch("time.sleep", immediate_return):
            yield

    else:
        yield
```

Comment (leti367) [reply sdk/devcenter/azure-developer-devcenter/tests/conftest.py:17]: Thanks @mccoyp, this solution works, tested locally. I just committed the changes in this PR. Thanks.

**Pull request merged**
